### ISSUE TITLE
Confusion about jvp input shapes

### ISSUE THREAD

**Issue by KeAWang** [google/jax#5226]
I have a question about `jax.jvp`. From my understanding, given `f: R^n -> R^m`, `jax.jvp(f, x, v)` will compute the pushforward at `x` evaluated on tangent vector `v`, i.e. `df(x)[v]`. In this case we can think about this as a matrix multiplicatoin where `df(x)` as a `m x n` matrix and `v` to be the same shape as `x`.

However, it seems that `jax.jvp` allow `v` to be of a different shape than `x` which confuses me:
```python
import jax
def f(x):
  # Note that f: R^n -> R^(n-1)
  return x[:-1]

n = 10
key = jax.random.PRNGKey(0)
x = jax.random.normal(key, (n,))
v = jax.random.normal(key, (2*n,))
jax.jvp(f, (x,), (v,))
```
will run properly and return
```python
(DeviceArray([-0.372111  ,  0.26423106, -0.18252774, -0.7368198 ,
              -0.44030386, -0.15214427, -0.6713536 , -0.5908642 ,
               0.73168874], dtype=float32),
 DeviceArray([ 1.054516  , -0.9692889 , -0.59460217, -0.0318858 ,
               2.410932  , -1.8784496 , -0.7847697 , -0.3137085 ,
               0.33370885], dtype=float32))
```
despite the fact that `v` is not the same shape as `x`. Shouldn't this fail since `x` and `v` are not the same shape?

**Comment by mattjj** [comment]
Great question!

It does indeed seem that we should add a check that `v` is the same shape as `x`. We have such a [check for `jax.linearize`](https://github.com/google/jax/blob/36f91261099b00194922bd93ed1286fe1c199724/jax/api.py#L1794-L1801) and [`jax.vjp`](https://github.com/google/jax/blob/36f91261099b00194922bd93ed1286fe1c199724/jax/api.py#L1815-L1821). We seem to have a [check for matching dtypes in `jax.jvp`](https://github.com/google/jax/blob/36f91261099b00194922bd93ed1286fe1c199724/jax/api.py#L1704-L1712); I wonder if there was some reason not to check the shapes as well...

**Comment by LuiGiovanni** [comment]
Hi, I will take this issue. It's my first time working on an issue here in Jax

**Comment by LuiGiovanni** [comment]
Hello @mattjj @KeAWang, I've been trying to replicate this issue and when I run `jax.vjp` I run into this error message:

![image](https://user-images.githubusercontent.com/16671869/103182471-bd9e5100-4868-11eb-916c-7fc73da65d98.png)

basically, telling me I have given f too many arguments, BUT if I run it inside a second pair a parenthesis I get the next output:

![image](https://user-images.githubusercontent.com/16671869/103182497-ee7e8600-4868-11eb-960c-4284b8192b27.png)

The highlighted pair a parenthesis is the ones I added, is this also an issue that must be checked? Is this the way to replicate the issue? Keep in mind I have tried this inside a jupyter notebook, though I get the same results in my local python interpreter. Thank you for everything.

**Comment by zhangqiaorjc** [comment]
Matt, i'm assigning to you to shepherd this issue

**Comment by gsp-27** [comment]
Hi,
while trying to fix this, I tried to also check the ```linearize``` function. I am confused by its output too.

```python
import jax

def f(x):
    return x+1

n = 10
key = jax.random.PRNGKey(0)
x = jax.random.normal(key, (n,))
v = jax.random.normal(key, (2*n,))

y, func = jax.linearize(f, x)
dd = func(v)
print(dd.shape)
print(dd)
```

This gives the output as
```
(20,)
[ 1.054516   -0.9692889  -0.59460217 -0.0318858   2.410932   -1.8784496
 -0.7847697  -0.3137085   0.33370885  1.767703   -1.0277647   1.4111717
 -0.5084972  -0.5263776   0.5031503   1.0549793  -0.0874074   0.79581654
  2.656559   -0.5822907 ]
```
Since, ```f``` is a function from $R^n \rightarrow R^n$, its jacobian would be a (10,10) matrix in this case, and computing the dot product with a (20,) dimensional vector should ideally return a shape-mismatch error.

I tried to dig a bit further and realized that https://github.com/google/jax/blob/ad3188fb7929e93f9500e7149631d9e35e6258bf/jax/core.py#L1029, gets executed.

Please correct me if I misunderstood something.

If it is indeed a problem, I would like to take up the issue.

Thanks,
Gaurav Pathak

**Comment by mkunesch** [comment]
This has been solved by #5446 (the provided code snippet now raises a `ValueError`). Please reopen if this is incorrect!
